proposals/multi-field-value-classes.md

```diff
@@ -0,0 +1,691 @@
+
+## Use cases
+*Since single field value classes are inline classes, which are already implemented, this KEEP describes further only multi-field value classes (MFVC).*
+
+The typical example of MFVC usage is applications that create a lot of small effective structures without identity, e.g. geometry structures, complex numbers, rational numbers:
+
+```kotlin
+interface Figure
+interface FigureWithPerimeter<T> : Figure {
+    val perimeter: T
+}
+interface FigureWithArea<T> : Figure {
+    val area: T
+}
+
+@JvmInline
+value class DPoint(val x: Double, val y: Double): Figure
+
+@JvmInline
+value class DSegment(val p1: DPoint, p2: DPoint): FigureWithPerimeter<Double> {
+    inline val middle: DPoint
+        get() = DPoint((p1.x + p2.x) / 2.0, (p1.y + p2.y) / 2.0)
+        
+    val length: Double
+        get() = sqrt((p1.x - p2.x).square() + (p1.y - p2.y).square())
+    
+    @TypedEquals
+    fun equals(other: DSegment) =
+        p1 == other.p1 && p2 == other.p2 || p1 == other.p2 && p2 == other.p1
+    
+    override fun hashCode() = p1.hashCode() + p2.hashCode()
+    
+    override val perimeter get() = length
+}
+
+@JvmInline
+value class DRectangle(
+    val topLeftVertex: DPoint, val bottomRightVertex: DPoint
+) : FigureWithPerimeter<Double>, FigureWithArea<Double> {
+    init {
+        require(topLeftVertex.x <= bottomRightVertex.x) {
+            "${topLeftVertex.x} > ${bottomRightVertex.x}"
+        }
+        require(topLeftVertex.y >= bottomRightVertex.y) {
+            "${topLeftVertex.y} < ${bottomRightVertex.y}"
+        }
+    }
+
+    inline val topRightVertex: DPoint
+        get() = DPoint(bottomRightVertex.x, topLeftVertex.y)
+    inline val bottomLeftVertex: DPoint
+        get() = DPoint(topLeftVertex.x, bottomRightVertex.y)
+    
+    val width: Double
+        get() = bottomRightVertex.x - topLeftVertex.x
+    
+    val hight: Double
+        get() = topLeftVertex.y - bottomRightVertex.y
+    
+    override val area: Double
+        get() = height * width
+    
+    override val perimeter: Double
+        get() = 2 * (height + width)
+}
+
+fun Double.square() = this * this
+```
+
+Another use case is creating some custom primitives for graphics:
+
+```kotlin
+@JvmInline
+value class Color(
+    val alpha: UByte, val red: UByte, val green: UByte, val blue: UByte
+)
+
+@JvmInline
+value class Gradient(val from: Color, val to: Color)
+
+fun Gradient.colorAt(percentage: Double): Color {
+     require(percentage in 0.0..100.0) { "Invalid percentage: $percentage" }
+     val c1 = 100.0 - percentage
+     val c2 = percentage
+     return Color(
+         alpha = (from.alpha.toDouble() * c1 + to.alpha.toDouble() * c2).toUByte(),
+         red = (from.red.toDouble() * c1 + to.red.toDouble() * c2).toUByte(),
+         green = (from.green.toDouble() * c1 + to.green.toDouble() * c2).toUByte(),
+         blue = (from.blue.toDouble() * c1 + to.blue.toDouble() * c2).toUByte(),
+     )
+}
```

Comment:
Looks like there is a logical mistake in this implementation... After executing this method, colors will be also multiplied by 100.

Probably additional division by 100 is needed in each line:

```suggestion
fun Gradient.colorAt(percentage: Double): Color {
     require(percentage in 0.0..100.0) { "Invalid percentage: $percentage" }
     val c1 = 100.0 - percentage
     val c2 = percentage
     return Color(
         alpha = (from.alpha.toDouble() * c1 + to.alpha.toDouble() * c2).div(100).toUByte(),
         red = (from.red.toDouble() * c1 + to.red.toDouble() * c2).div(100).toUByte(),
         green = (from.green.toDouble() * c1 + to.green.toDouble() * c2).div(100).toUByte(),
         blue = (from.blue.toDouble() * c1 + to.blue.toDouble() * c2).div(100).toUByte(),
     )
}
```

Or `0.0..1.0` range should be used:

```suggestion
fun Gradient.colorAt(ratio: Double): Color {
     require(ratio in 0.0..1.0) { "Invalid ratio: $percentage" }
     val c1 = 1.0 - ratio
     val c2 = ratio
     return Color(
         alpha = (from.alpha.toDouble() * c1 + to.alpha.toDouble() * c2).toUByte(),
         red = (from.red.toDouble() * c1 + to.red.toDouble() * c2).toUByte(),
         green = (from.green.toDouble() * c1 + to.green.toDouble() * c2).toUByte(),
         blue = (from.blue.toDouble() * c1 + to.blue.toDouble() * c2).toUByte(),
     )
}
```

Reply:
I would go for ratio too. Minor typo:

```
require(ratio in 0.0..1.0) { "Invalid ratio: $percentage" }
```
->
```
require(ratio in 0.0..1.0) { "Invalid ratio: $ratio" }
```